### ISSUE TITLE
The gradient changes after concatenating in pytorch

### ISSUE THREAD

**Issue by zhiheLu** [pytorch/pytorch#25642]
## Issue description

I found after concatenated the gradient of the input is different. Could you help me find why? Many thanks in advance.

## Code example
```
import torch
a = torch.tensor([1.,2.], requires_grad=True)
b = torch.tensor([2.,3.], requires_grad=True)  
ab = torch.cat((a,b), dim=0) # ab  = tensor([1., 2., 2., 3.], grad_fn=<CatBackward>)
z = ab**2 # z = tensor([1., 4., 4., 9.], grad_fn=<PowBackward0>)
out = z.mean() # out = tensor(4.5000, grad_fn=<MeanBackward0>)
out.backward()
a.grad # tensor([1.5000, 2.5000])
b.grad # tensor([1.5000, 2.5000])
```

But when I compute the gradients by hand, the results are as follows:
```
a = [x1, x2]
b = [x3, x4]
ab = [x1, x2, x3, x4]
z = [x1^2, x2^2, x3^2, x4^2]
out = 1/4(x1^2 + x2^2 + x3^2 + x4^2)
So, use out to compute the gradient for a should be: [1/2(x1), 1/2(x2)] = [0.5, 1.0]
as well, use out to compute the gradient for b should be: [1/2(x3), 1/2(x4)] = [1.0, 1.5]
```

- PyTorch:
- PyTorch version: '1.2.0'
- Python version: '3.7.4'

**Comment by soumith** [comment]
please use https://discuss.pytorch.org for questions
